File: test/create.test.js
const { describe, it } = require("node:test");
const assert = require("node:assert/strict");
const fs = require("fs");
const path = require("path");
const childProcess = require("child_process");
const gitModule = require("../lib/git");
const configModule = require("../lib/config");

const ROOT = "/projects/myrepo";

function setup(t, configOverrides = {}) {
  const config = {
    remote: "origin",
    defaultBase: "main",
    copyFiles: [],
    postCreateScript: null,
    ...configOverrides,
  };

  t.mock.method(configModule, "loadConfig", () => ({ root: ROOT, config }));
  t.mock.method(gitModule, "gitInBare", (args) => {
    if (args.includes("rev-parse")) throw new Error("not found");
    return "";
  });
  t.mock.method(fs, "existsSync", () => false);
  t.mock.method(fs, "cpSync", () => {});
  t.mock.method(fs, "mkdirSync", () => {});
  t.mock.method(childProcess, "execSync", () => "");
  t.mock.method(console, "log", () => {});
  t.mock.method(console, "error", () => {});
  t.mock.method(process, "exit", (code) => {
    throw new Error(`exit(${code})`);
  });

  // Require module BEFORE mocking readFileSync/writeFileSync
  // (Node's require uses fs.readFileSync internally to load .js files)
  delete require.cache[require.resolve("../lib/commands/create")];
  const create = require("../lib/commands/create");

  t.mock.method(fs, "readFileSync", (p) => {
    if (String(p).endsWith(".git")) return "gitdir: /fake/gitdir";
    return "";
  });
  t.mock.method(fs, "writeFileSync", () => {});

  return create;
}

describe("create", () => {
  it("fetches, creates worktree, and sets upstream", (t) => {
    const create = setup(t);
    create("my-branch", {});

    const calls = gitModule.gitInBare.mock.calls.map((c) => c.arguments[0]);
    assert.equal(calls[0], "fetch origin");
    assert.ok(calls[2].includes("worktree add"));
    assert.ok(calls[2].includes(path.join(ROOT, "my-branch")));
    assert.ok(calls[2].includes("-b my-branch"));
    assert.ok(calls[2].includes("origin/main"));
  });

  it("uses custom base branch from opts", (t) => {
    const create = setup(t);
    create("my-branch", { base: "develop" });

    const calls = gitModule.gitInBare.mock.calls.map((c) => c.arguments[0]);
    assert.ok(calls[2].includes("origin/develop"));
  });

  it("exits if branch directory already exists", (t) => {
    const create = setup(t);
    fs.existsSync.mock.mockImplementation(() => true);

    assert.throws(() => create("my-branch", {}), { message: "exit(1)" });
  });

  it("copies files that exist at source", (t) => {
    const create = setup(t, { copyFiles: [".env", "config.json"] });
    fs.existsSync.mock.mockImplementation((p) => {
      if (p === path.join(ROOT, "my-branch")) return false;
      return true;
    });

    create("my-branch", {});

    assert.equal(fs.cpSync.mock.callCount(), 2);
    // Source is base worktree (main) since it exists
    assert.equal(
      fs.cpSync.mock.calls[0].arguments[0],
      path.join(ROOT, "main", ".env")
    );
    assert.equal(
      fs.cpSync.mock.calls[0].arguments[1],
      path.join(ROOT, "my-branch", ".env")
    );
  });

  it("skips missing copy files silently", (t) => {
    const create = setup(t, { copyFiles: [".env"] });
    // existsSync returns false for everything (default)
    create("my-branch", {});
    assert.equal(fs.cpSync.mock.callCount(), 0);
  });

  it("runs post-create script in worktree dir", (t) => {
    const create = setup(t, { postCreateScript: "npm install" });
    create("my-branch", {});

    assert.equal(childProcess.execSync.mock.callCount(), 1);
    assert.equal(
      childProcess.execSync.mock.calls[0].arguments[0],
      "npm install"
    );
    assert.equal(
      childProcess.execSync.mock.calls[0].arguments[1].cwd,
      path.join(ROOT, "my-branch")
    );
  });

  it("skips post-create script when not configured", (t) => {
    const create = setup(t);
    create("my-branch", {});
    assert.equal(childProcess.execSync.mock.callCount(), 0);
  });

  it("writes worktree config to disable bare mode", (t) => {
    const create = setup(t);
    create("my-branch", {});

    const writeCalls = fs.writeFileSync.mock.calls;
    const configCall = writeCalls.find((c) =>
      String(c.arguments[0]).includes("config.worktree")
    );
    assert.ok(configCall, "should write config.worktree");
    assert.ok(configCall.arguments[1].includes("bare = false"));
    assert.ok(configCall.arguments[1].includes("autoSetupRemote = true"));
  });

  it("exports setupWorktree", (t) => {
    const create = setup(t);
    assert.equal(typeof create.setupWorktree, "function");
  });
});

File: internal/cmd/create_test.go
package cmd

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/gradyholmes/wt/internal/config"
)

func TestCreateCmd(t *testing.T) {
	t.Run("fetches, creates worktree, and sets upstream", func(t *testing.T) {
		root := t.TempDir()
		mg := &MockGit{
			GitInBareFunc: func(args, projectRoot string) (string, error) {
				if strings.Contains(args, "rev-parse") {
					return "", fmt.Errorf("not found")
				}
				// For worktree add, create the worktree dir and .git file
				if strings.Contains(args, "worktree add") {
					wtPath := filepath.Join(root, "my-branch")
					os.MkdirAll(wtPath, 0755)
					gitDir := filepath.Join(root, ".bare", "worktrees", "my-branch")
					os.MkdirAll(gitDir, 0755)
					os.WriteFile(filepath.Join(wtPath, ".git"), []byte("gitdir: "+gitDir), 0644)
				}
				return "", nil
			},
		}
		mc := &MockConfig{Root: root, Cfg: &config.Config{
			Remote:      "origin",
			DefaultBase: "main",
		}}
		d, _, _ := testDeps(mg, mc, nil)

		cmd := CreateCmd(d)
		cmd.SilenceUsage = true
		cmd.SetArgs([]string{"my-branch"})
		if err := cmd.Execute(); err != nil {
			t.Fatal(err)
		}

		calls := mg.Calls
		if len(calls) < 3 {
			t.Fatalf("expected at least 3 git calls, got %d: %v", len(calls), calls)
		}
		if calls[0] != "fetch origin" {
			t.Errorf("first call = %q, want fetch origin", calls[0])
		}
		addCall := calls[2]
		if !strings.Contains(addCall, "worktree add") {
			t.Errorf("third call should be worktree add, got %q", addCall)
		}
		if !strings.Contains(addCall, "-b my-branch") {
			t.Errorf("should create branch my-branch, got %q", addCall)
		}
		if !strings.Contains(addCall, "origin/main") {
			t.Errorf("should be based on origin/main, got %q", addCall)
		}
	})

	t.Run("uses custom base branch from --base flag", func(t *testing.T) {
		root := t.TempDir()
		mg := &MockGit{
			GitInBareFunc: func(args, projectRoot string) (string, error) {
				if strings.Contains(args, "rev-parse") {
					return "", fmt.Errorf("not found")
				}
				if strings.Contains(args, "worktree add") {
					wtPath := filepath.Join(root, "my-branch")
					os.MkdirAll(wtPath, 0755)
					gitDir := filepath.Join(root, ".bare", "worktrees", "my-branch")
					os.MkdirAll(gitDir, 0755)
					os.WriteFile(filepath.Join(wtPath, ".git"), []byte("gitdir: "+gitDir), 0644)
				}
				return "", nil
			},
		}
		mc := &MockConfig{Root: root, Cfg: &config.Config{
			Remote:      "origin",
			DefaultBase: "main",
		}}
		d, _, _ := testDeps(mg, mc, nil)

		cmd := CreateCmd(d)
		cmd.SilenceUsage = true
		cmd.SetArgs([]string{"my-branch", "--base", "develop"})
		if err := cmd.Execute(); err != nil {
			t.Fatal(err)
		}

		addCall := mg.Calls[2]
		if !strings.Contains(addCall, "origin/develop") {
			t.Errorf("should use develop base, got %q", addCall)
		}
	})

	t.Run("errors if branch directory already exists", func(t *testing.T) {
		root := t.TempDir()
		os.MkdirAll(filepath.Join(root, "my-branch"), 0755)

		mg := &MockGit{}
		mc := &MockConfig{Root: root, Cfg: &config.Config{
			Remote:      "origin",
			DefaultBase: "main",
		}}
		d, _, _ := testDeps(mg, mc, nil)

		cmd := CreateCmd(d)
		cmd.SilenceErrors = true
		cmd.SilenceUsage = true
		cmd.SetArgs([]string{"my-branch"})
		err := cmd.Execute()
		if err == nil {
			t.Fatal("expected error")
		}
		if !strings.Contains(err.Error(), "already exists") {
			t.Errorf("error = %q", err)
		}
	})

	t.Run("writes worktree config to disable bare mode", func(t *testing.T) {
		root := t.TempDir()
		mg := &MockGit{
			GitInBareFunc: func(args, projectRoot string) (string, error) {
				if strings.Contains(args, "rev-parse") {
					return "", fmt.Errorf("not found")
				}
				if strings.Contains(args, "worktree add") {
					wtPath := filepath.Join(root, "my-branch")
					os.MkdirAll(wtPath, 0755)
					gitDir := filepath.Join(root, ".bare", "worktrees", "my-branch")
					os.MkdirAll(gitDir, 0755)
					os.WriteFile(filepath.Join(wtPath, ".git"), []byte("gitdir: "+gitDir), 0644)
				}
				return "", nil
			},
		}
		mc := &MockConfig{Root: root, Cfg: &config.Config{
			Remote:      "origin",
			DefaultBase: "main",
		}}
		d, _, _ := testDeps(mg, mc, nil)

		cmd := CreateCmd(d)
		cmd.SilenceUsage = true
		cmd.SetArgs([]string{"my-branch"})
		if err := cmd.Execute(); err != nil {
			t.Fatal(err)
		}

		configWorktree := filepath.Join(root, ".bare", "worktrees", "my-branch", "config.worktree")
		data, err := os.ReadFile(configWorktree)
		if err != nil {
			t.Fatal("config.worktree should be written")
		}
		if !strings.Contains(string(data), "bare = false") {
			t.Error("should contain bare = false")
		}
		if !strings.Contains(string(data), "autoSetupRemote = true") {
			t.Error("should contain autoSetupRemote = true")
		}
	})

	t.Run("copies files that exist at source", func(t *testing.T) {
		root := t.TempDir()
		// Create base worktree with files to copy
		os.MkdirAll(filepath.Join(root, "main"), 0755)
		os.WriteFile(filepath.Join(root, "main", ".env"), []byte("SECRET=1"), 0644)
		os.WriteFile(filepath.Join(root, "main", "config.json"), []byte("{}"), 0644)

		mg := &MockGit{
			GitInBareFunc: func(args, projectRoot string) (string, error) {
				if strings.Contains(args, "rev-parse") {
					return "", fmt.Errorf("not found")
				}
				if strings.Contains(args, "worktree add") {
					wtPath := filepath.Join(root, "my-branch")
					os.MkdirAll(wtPath, 0755)
					gitDir := filepath.Join(root, ".bare", "worktrees", "my-branch")
					os.MkdirAll(gitDir, 0755)
					os.WriteFile(filepath.Join(wtPath, ".git"), []byte("gitdir: "+gitDir), 0644)
				}
				return "", nil
			},
		}
		mc := &MockConfig{Root: root, Cfg: &config.Config{
			Remote:      "origin",
			DefaultBase: "main",
			CopyFiles:   []string{".env", "config.json"},
		}}
		d, stdout, _ := testDeps(mg, mc, nil)

		cmd := CreateCmd(d)
		cmd.SilenceUsage = true
		cmd.SetArgs([]string{"my-branch"})
		if err := cmd.Execute(); err != nil {
			t.Fatal(err)
		}

		// Verify files were copied
		data, err := os.ReadFile(filepath.Join(root, "my-branch", ".env"))
		if err != nil {
			t.Fatal("should copy .env")
		}
		if string(data) != "SECRET=1" {
			t.Errorf(".env content = %q", data)
		}

		if !strings.Contains(stdout.String(), "Copied .env") {
			t.Error("should log copied files")
		}
	})

	t.Run("skips missing copy files silently", func(t *testing.T) {
		root := t.TempDir()
		mg := &MockGit{
			GitInBareFunc: func(args, projectRoot string) (string, error) {
				if strings.Contains(args, "rev-parse") {
					return "", fmt.Errorf("not found")
				}
				if strings.Contains(args, "worktree add") {
					wtPath := filepath.Join(root, "my-branch")
					os.MkdirAll(wtPath, 0755)
					gitDir := filepath.Join(root, ".bare", "worktrees", "my-branch")
					os.MkdirAll(gitDir, 0755)
					os.WriteFile(filepath.Join(wtPath, ".git"), []byte("gitdir: "+gitDir), 0644)
				}
				return "", nil
			},
		}
		mc := &MockConfig{Root: root, Cfg: &config.Config{
			Remote:      "origin",
			DefaultBase: "main",
			CopyFiles:   []string{".env"},
		}}
		d, stdout, _ := testDeps(mg, mc, nil)

		cmd := CreateCmd(d)
		cmd.SilenceUsage = true
		cmd.SetArgs([]string{"my-branch"})
		if err := cmd.Execute(); err != nil {
			t.Fatal(err)
		}

		if strings.Contains(stdout.String(), "Copied") {
			t.Error("should not log copies for missing files")
		}
	})
}
